test/functional/p2p_invalid_messages.py

```diff
@@ -42,13 +47,32 @@ def set_test_params(self):
         self.setup_clean_chain = True
 
     def run_test(self):
+        self.test_buffer()
         self.test_magic_bytes()
         self.test_checksum()
         self.test_size()
         self.test_msgtype()
         self.test_large_inv()
         self.test_resource_exhaustion()
 
+    def test_buffer(self):
+        self.log.info("Test message with header split across two buffers, should be received")
+        conn = self.nodes[0].add_p2p_connection(P2PDataStore())
+        # Create valid message
+        msg = conn.build_message(msg_ping(nonce=12345))
+        cut_pos = 12    # Chosen at an arbitrary position within the header
+        # Send message in two pieces
+        before = int(self.nodes[0].getnettotals()['totalbytesrecv'])
+        conn.send_raw_message(msg[:cut_pos])
+        time.sleep(0.01) # Prevent the RPC call from going too early
```

Comment:
Actually, I think `wait_until` might be better:
```suggestion
       wait_until(lambda: self.nodes[0].getnettotals()['totalbytesrecv'] == before + cut_pos)
```
Wishful thinking, it'd be nice if `cut_pos` is a size that couldn't possibly be = another message (does this exist?). It will still fail if the node receives another message in between the halves, but can't pass if it isn't received.

Reply:
Ah we came up with almost the same thing!  I thought of this https://github.com/bitcoin/bitcoin/pull/19304#discussion_r441746483.  Regarding `cut_pos`, we actually have what you're looking for.  `cut_pos = 12` and the size of the smallest message (any of the messages with no payload) is 24 😄 

I'll think about `getpeerinfo` vs `getnettotals`

Reply:
>I don't think this guarantees that conn's message has been processed just because pinger's ping has been though...?

I'm pretty sure you're right.  We're looking at a race in a test for a race inside a test at this point, I'm going to prefer the simpler fix.

Reply:
TWINSIES 👯 
(If I may, I think `== before + cut_pos` is stronger than  `>= before` or `!= before`)
> we actually have what you're looking for

Ooh!! Genius, should be in the comment.

Reply:
`== before + cut_pos` is stronger, but I see three possible situations that I would rather be caught by different asserts.  This is possible with `!=` and the later `assert_equal`

1. Something terrible happens and the first half of the message is never processed.  `wait_until` times out and throws an exception.
2. The node takes a while but eventually processes the first half of the message (and only that).  `wait_until` tries `getnettotals` a few times and then we continue past, and the test passes.
3.  The race occurs and the node processes more than just the first half of the message.  The `assert_equals` fails and we see the exception there.

Choosing `== before + cut_pos` or `>= before` could make both 1. and 3. fail on the `wait_until`, which is just a little harder to debug.